api/types/databaseobject/databaseobject.go

```diff
@@ -0,0 +1,75 @@
+// Copyright 2024 Gravitational, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package databaseobject
+
+import (
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/defaults"
+	dbobjectv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/dbobject/v1"
+	headerv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/header/v1"
+	"github.com/gravitational/teleport/api/types"
+)
+
+// NewDatabaseObject creates a new dbobjectv1.DatabaseObject.
+func NewDatabaseObject(name string, spec *dbobjectv1.DatabaseObjectSpec) (*dbobjectv1.DatabaseObject, error) {
+	return NewDatabaseObjectWithLabels(name, nil, spec)
+}
+
+// NewDatabaseObjectWithLabels creates a new dbobjectv1.DatabaseObject with specified labels.
+func NewDatabaseObjectWithLabels(name string, labels map[string]string, spec *dbobjectv1.DatabaseObjectSpec) (*dbobjectv1.DatabaseObject, error) {
+	databaseObject := &dbobjectv1.DatabaseObject{
+		Kind:    types.KindDatabaseObject,
+		Version: types.V1,
+		Metadata: &headerv1.Metadata{
+			Name:      name,
+			Namespace: defaults.Namespace,
+			Labels:    labels,
+		},
+		Spec: spec,
+	}
+
+	err := ValidateDatabaseObject(databaseObject)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return databaseObject, nil
+}
+
+// ValidateDatabaseObject checks if *dbobjectv1.DatabaseObject is valid.
+func ValidateDatabaseObject(obj *dbobjectv1.DatabaseObject) error {
```

Comment:
Any advantage to having this `package.ValidateXXX` function vs the common `obj.CheckAndSetDefaults`? Do we expect the function to be used other places?

Reply:
We can't have `CheckAndSetDefaults()` or any other custom methods in the RFD 153 world.

Reply:
I didn't catch that part. Let me read through it again =p

Reply:
I'm not sure if this is explicitly mentioned in the RFD, but that is the reality right now. We could, in theory, write a custom codegen for subset of our message types, but there is a resistance to this notion: codegens are somewhat hard to maintain.